bpo-31908: Fix output of cover files for trace module command-line tool

selik opened the pull request:
The `lnotab` variable -- a dict of executable lines -- is a helper for the `show_missing` parameter. When `show_missing` is false, we don't need to calculate that dict. However, emitting the cover files had been made conditional on that dict being non-empty in Commit f026dae130bf6f9015c4b212f16852ba4a3f3dec. This reverts part of the commit to ensure the cover files are emitted regardless of the `show_missing` argument.

Additionally, the `write_results_file` is refactored to clarify the purpose of `lnotab`.

<!-- issue-number: bpo-31908 -->
https://bugs.python.org/issue31908
<!-- /issue-number -->

selik committed:
flatten if/elif/else for line annotation

Rather than nesting an if/else inside the last else of the outer
condition, the conditional can be flattened into four possibilities.

selik committed:
reorder condition for line annotation

A blank line will not appear in ``lnotab`` as returned by
``_find_executable_linenos``. Alternately, ``lnotab`` might be set to an
empty dict. Either way, the check for whether a line is blank can be
done after checking for whether the ``lineno`` is in ``lnotab``.

selik committed:
delete redundant condition

There's no need to check for an empty line. Either we hit the line and
wrote the count, we didn't hit the line and wrote an annotation, or we
want to print some spaces to indent the line.

selik committed:
remove unused regex

selik committed:
explain (comment) lnotab parameter

The lnotab parameter shouldn't be renamed, for compatibility, but it
could use some further explanation. One natural reading of it would be
"lines without tabs", but it's really a "line number table" -- a mapping
of line numbers to possible executable lines.

One nuance to that "table" is that ``_find_executable_linenos`` will
ignore some code that the user might want to highlight as unreachable.
For example, code inside a condition like ``if False:`` will be
excluded.

Further, it appears ``lnotab`` doesn't need to be a dict. A set would
suffice and might reduce the memory requirements of the trace module. If
the ``_find_lines`` function is used only by the trace module, as the
underscore prefix suggests, it could be refactored to return a set of
line numbers instead of a mapping of line numbers to lines.

selik committed:
write cover files regardless of show_missing

Whether the show_missing parameter is true or false, the cover file
should be written. The ``lnotab`` parameter will be used simply to
decide whether to annotate unreached lines with some extra characters.

selik committed:
add news blurb

selik committed:
Add testcase for verifying CLI output

The single test function checks that the coverfile is correctly written
and contains appropriate line count and unreached code highlighting.

The check for ``if []`` avoids interpreter optimization which might
allow the interpreter to recognize the conditional statement as a no-op
and eliminate it. For example, ``if False`` would not be counted, nor
would any code in the following block be highlighted by the trace.

serhiy-storchaka commented on Lib/test/test_trace.py:
```diff
@@ -365,6 +366,35 @@ def test_ignored(self):
+    def setUp(self):
+        with open(self.codefile, 'w') as f:
+            f.write(textwrap.dedent('''\
+            x = 42
```
Please increase the indentation of these lines for readability.

serhiy-storchaka reviewed (commented):
The test is passed if revert changes in `trace.py`. Please ensure that it fails with unpatched code.

selik committed:
Add test without --missing CLI option

The --missing or -m option was hiding the bug. Testing the CLI with
--count option and only that option reveals the bug.

selik committed:
fix test expected output indentation
